### PR TITLE
Provide return type of ReflectionParameter::getType()

### LINKED ISSUE #5258: ReflectionParameter::hasType() should assert return type of ReflectionParameter::getType()
If `ReflectionParameter::hasType()` returns `true`, then `ReflectionParameter::getType()` will always return `ReflectionType`.

If `ReflectionParameter::hasType()` returns `false`, then `ReflectionParameter::getType()` will always return `null`.

This is not currently understood by psalm.

https://psalm.dev/r/257f4e884d

### PR DESCRIPTION
**Pull request by dbrekelmans**
Fixes #5258

Thanks to @orklah for giving me some hints 👍

### CODE REVIEW

**Review by nusje2000 — APPROVED** [review]

**Comment by weirdan** [comment]
@dbrekelmans mind fixing CS (https://circleci.com/gh/vimeo/psalm/22383)?

Tests failures are unrelated, fixed in #5335

File: src/Psalm/Internal/Provider/ReturnTypeProvider/ReflectionParameterGetType.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+        $scopedVarName = '$' . $stmt->var->name . '->hastype()';
+        if (!isset($event->getContext()->vars_in_scope[$scopedVarName])) {
+            return null;
+        }
+
+        $type = $event->getContext()->vars_in_scope[$scopedVarName];
+
+        if ($type->isTrue()) {
+            return Type::parseString(ReflectionType::class);
```
Comment (weirdan) [inline src/Psalm/Internal/Provider/ReturnTypeProvider/ReflectionParameterGetType.php:46]:
```suggestion
            return new Type\Union([
                new Type\Atomic\TNamedObject(ReflectionType::class)
            ]);
```
There's no need to go through parser as we know exactly what we need to return.

**Review by weirdan — CHANGES_REQUESTED** [review]

File: src/Psalm/Internal/Provider/ReturnTypeProvider/ReflectionParameterGetType.php (hunk 1 above)
Comment (dbrekelmans) [reply src/Psalm/Internal/Provider/ReturnTypeProvider/ReflectionParameterGetType.php:46]: Ah yes, I tried initially, but couldn't find out how to do it. Thanks!

**Review by weirdan — APPROVED** [review]

**Comment by muglug** [comment]
First, thanks for this!

I think I prefer the option you presented in the original issue, following [this recent dictum](https://psalm.dev/docs/contributing/philosophy/#docblock-annotations-are-better-than-type-providing-plugins).

Given that `ReflectionParameter` is an immutable class the assertion would be entirely safe (though it requires a bugfix for Psalm to allow this specific syntax, which I'm working on now.

```php
<?php

/**
 * @immutable
 */
class ReflectionParameter {
  /**
   * @psalm-assert-if-true $this->getType() ReflectionType
   * @psalm-assert-if-false $this->getType() null
   */
  public function hasType() : bool {
      return true;
  }
    
  public function getType() : ?ReflectionType {
   	return null;   
  }
}
```

The PR is not for nought, as I'll re-use your tests.

**Comment by orklah** [comment]
@muglug I'm kinda lost
The else part seems to correctly infer that the type is null: https://psalm.dev/r/82732b730e

However, there is no assert-if-false in the stub like in your example above:
https://github.com/vimeo/psalm/blob/master/stubs/Reflection.phpstub#L108

How does that works then?

**Comment by weirdan** [comment]
> How does that works then?

I mentioned it here: https://github.com/vimeo/psalm/issues/5248#issuecomment-781339212
Looks like a bug to me.

**Pull request closed without merging**